### DISCUSSION TITLE
Upgrading from 2.4.2 to 2.4.7

### DISCUSSION THREAD

**Discussion by dynamicapps** [markshust/docker-magento#1197] [Q&A] [answered]
I have my 2.4.2 site from cloud commerce running in docker (thanks Mark!). I want to upgrade to 2.4.7 in a docker instance before I do it on my staging and then production servers.
My main question is which containers I should just get current images of vs. which I need to go through an upgrade process on.
I appreciate any suggestions.

**Comment by YevhenZvieriev** [comment, marked as answer]
Hello, @dynamicapps 

I have also upgraded my sandbox Magento instance from 2.4.2 to 2.4.7 so hope I will help you :)

## Containers to Simply Update:

### Nginx (app):

2.4.2: markoshust/magento-nginx:1.18-8

2.4.7: markoshust/magento-nginx:1.24-0

### PHP-FPM (phpfpm):

2.4.2: markoshust/magento-php:7.4-fpm

2.4.7: markoshust/magento-php:8.3-fpm-2

### Redis:

2.4.2: redis:6.0

2.4.7: redis:7.2-alpine

### Mailcatcher:

Image hasn't changed: sj26/mailcatcher

No update is needed.

### RabbitMQ:

2.4.2: rabbitmq:3.8

2.4.7: markoshust/magento-rabbitmq:3.12-0

## Containers Requiring Additional Steps:

### MariaDB (db):

2.4.2: mariadb:10.4

2.4.7: mariadb:10.6

Updating the database may require additional migration steps. It is recommended to back up your data before updating.

### Elasticsearch/OpenSearch:

2.4.2: markoshust/magento-elasticsearch:7.9

2.4.7: markoshust/magento-opensearch:2.12-0

After updating the image, you might need to update the Elasticsearch or OpenSearch indices.

In my case `compose.yaml` for 2.4.7 looks like:

``` 
services:
  app:
    image: markoshust/magento-nginx:1.24-0
    ports:
      - "80:8000"
      - "443:8443"
    volumes: &appvolumes
      - ~/.composer:/var/www/.composer:cached
      - ~/.ssh/id_rsa:/var/www/.ssh/id_rsa:cached
      - ~/.ssh/known_hosts:/var/www/.ssh/known_hosts:cached
      - appdata:/var/www/html
      - sockdata:/sock
      - ssldata:/etc/nginx/certs

  phpfpm:
    image: markoshust/magento-php:8.3-fpm-2
    volumes: *appvolumes
    env_file: env/phpfpm.env

  db:
    image: mariadb:10.6
    command:
      --max_allowed_packet=64M
      --optimizer_use_condition_selectivity=1
      --optimizer_switch="rowid_filter=off"
    ports:
      - "3306:3306"
    env_file: env/db.env
    volumes:
      - dbdata:/var/lib/mysql

  redis:
    image: redis:7.2-alpine
    ports:
      - "6379:6379"

  opensearch:
    image: markoshust/magento-opensearch:2.12-0
    ports:
      - "9200:9200"
      - "9300:9300"
    env_file: env/opensearch.env
    environment:
      - "discovery.type=single-node"
      - "cluster.routing.allocation.disk.threshold_enabled=false"
      - "index.blocks.read_only_allow_delete"

  rabbitmq:
    image: markoshust/magento-rabbitmq:3.12-0
    ports:
      - "15672:15672"
      - "5672:5672"
    volumes:
      - rabbitmqdata:/var/lib/rabbitmq
    env_file: env/rabbitmq.env

  mailcatcher:
    image: sj26/mailcatcher
    ports:
      - "1080:1080"
    
volumes:
  appdata:
  dbdata:
  rabbitmqdata:
  sockdata:
  ssldata:
```
## My friendly recommendations:

1. Always create backups of important data, particularly databases, before updating.

2. Test the update on a staging environment before making changes in production.

3. Review the official Magento documentation and release notes to account for all possible changes and requirements for the new version.

BTW, feel free to use the script `bin/check-dependencies` that I developed, hope it will be useful for you during the upgrading process.

If you have any additional questions or need further assistance, feel free to reach out :)

**Reply by dynamicapps** [reply]
So, as I understand it, I should update all of my service containers first and then I follow the same steps that I did when setting up my current docker environment, correct? 
Here's what I'm not sure about:
I've run the upgrade compatibility tool and it has pointed out some changes, mostly having to do with the php upgrade. Do I make any changes to my existing code before setting it up in the php8 container (where it will not run). Or do I just set that up and then start doing the upgrade (upgrading composer, updating the metapackage, upgrading my 3rd party products, fixing any dependency issues), then just start running uct again against the new server, fixing problems until I can get it to run?
Thanks!

**Reply by YevhenZvieriev** [reply]
Hi, @dynamicapps 

Yes, you should update all your service containers first. After updating the service containers, follow the same steps you did when setting up your current Docker environment.

Regarding your specific questions about the PHP upgrade and code compatibility:

### Upgrade PHP Environment:

First, switch to the PHP 8.3 container (`markoshust/magento-php:8.3-fpm-2`) and make sure your environment is running. This step ensures that you are working within the new PHP version context from the beginning.

### Code Compatibility Check:

Before running the upgrade, you should review and update your existing code for PHP 8 compatibility. The upgrade compatibility tool you mentioned should help identify areas that need changes. Fix these issues in your codebase to ensure it runs in the PHP 8 container.

### Upgrade Process:

Once your code is compatible with PHP 8, proceed with the Magento upgrade:

- Update Composer dependencies and the Magento metapackage to 2.4.7.

- Upgrade any third-party modules and extensions to versions compatible with Magento 2.4.7 and PHP 8.

- Resolve any dependency issues that arise during this process.

### Testing and Fixing Issues:

After upgrading your Magento instance and dependencies, start running the Upgrade Compatibility Tool (UCT) against your new setup. Fix any issues the tool identifies until your application runs smoothly.

### General Tips:

- Always keep backups of your data and configurations before making significant changes.

- Use a staging environment to test the upgrade process thoroughly before applying changes to your production environment.

- Refer to the official Magento documentation and release notes for any additional requirements or changes specific to the version you are upgrading to.

If you need any further assistance or run into specific issues during the upgrade, feel free to reach out. Good luck with your upgrade!

**Reply by dynamicapps** [reply]
Thanks for the additional assistance. Adobe docs are big on details, not so good on overviews.
John

**Reply by dynamicapps** [reply]
So this has gone well up to this point. All containers are updated. 
My composer require section is:
        "magento/magento-cloud-metapackage": "^2.4.7",
        "magento/magento-cloud-patches": "^1.0.27",
        "cweagans/composer-patches": "^1.7",
        "google/apiclient": "^2.0",
        "liquid/liquid": "^1.4",
        "magento/composer-root-update-plugin": "^2.0.3",
        "magento/extension-b2b": "^1.4"
 When I try to run magento in my container, I get "
 Circular dependency: Magento\Framework\App\Cache\Type\Config depends on Magento\Framework\Interception\PluginList\PluginList and vice versa."
 I'm at a loss since this is in the framework and they don't even appear to reference each other.
 I've tried clearing caches and removing files, etc., but have not been able to get past this.
 Solution or troubleshooting ideas?
 Thanks!
John

**Reply by YevhenZvieriev** [reply]
Hi, @dynamicapps 

It sounds like you've made good progress so far. The issue you're encountering, a circular dependency in the Magento framework, can be tricky.

Have you cleaned all caches using the `bin/magento c:c` and `bin/magento c:f` and the `generated` folder before running `setup:upgrade`?

**Reply by dynamicapps** [reply]
Yes, but I just flushed the redis cache and that seems to have gotten me past that.
Now I can't create a folder on the php container in /run with permissions failure, but the folder permissions seem ok.
Thanks.
-jpr

**Reply by YevhenZvieriev** [reply]
You are right, I was just about to tell you about the Redis
I found this [commentary](https://github.com/magento/magento2/issues/31183#issuecomment-897807685) that looks quite useful.

To fix permissions for the folders there are the `bin/fixowns` and `bin/fixperms` scripts. Could you try to use each of them, and tell me the result?

**Reply by dynamicapps** [reply]
Yes, they work fine, but I am still unable to create that folder in /run.  I think this is necessary because I am using a cloud commerce instance. Is there a password to use with su in the php container?

**Reply by dynamicapps** [reply]
Answering my question here. I could create it by doing docker exec -u 0 -it magentodockerdev-phpfpm-1 bash.

**Reply by YevhenZvieriev** [reply]
You are right, but I would recommend you to use `bin/bash` which calls `bin/cli`, which in turn calls `docker-compose exec`. This is an additional layer of abstraction that allows you to use commands from docker-compose rather than directly from Docker, `docker-compose exec` by default uses the user running inside the container (usually not root).

If you want to run any CLI command as root within the container feel free to use the `bin/root` or  `bin/rootnotty` scripts.

In summary, `bin/bash` uses a more complex abstraction to work with docker-compose, which can be useful for automating and managing environments. At the same time, `docker exec` gives you direct access to a specific container.

**Comment by dynamicapps** [comment]
@YevhenZvieriev,
You're a rock star, thanks. I knew people have done this before and hoped someone took good notes :-).
Thanks so much for taking the time to post.
John
